Update Access Control docs (#719)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/access.adoc

:ownable-cairo: link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-0.7.0/src/access/ownable/ownable.cairo[Ownable]
:sn_keccak: https://docs.starknet.io/documentation/architecture_and_concepts/Cryptography/hash-functions/#starknet_keccak[sn_keccak]
:extensibility-pattern: xref:extensibility.adoc#the_pattern

= Access

> Access control--that is, "who is allowed to do this thing"--is incredibly important in the world of smart contracts.

Reviewer: ```suggestion
Access control--that is, "who is allowed to do this thing"—is incredibly important in the world of smart contracts.
```
It's displayed as two dashes instead of an em dash

Some ideas to consider:

* Use {sn_keccak} instead.
* Use Cairo friendly hashing algorithms like Poseidon, which are implemented in the
https://github.com/starkware-libs/cairo/blob/main/corelib/src/poseidon.cairo[Cairo corelib].

TIP: The `selector!` macro can be used to compute {sn_keccak} in Cairo.

=== Interface

This is the full interface of the `AccessControl` implementation:

[,javascript]
----
trait IAccessControl {
    /// Returns whether the account has the role or not.
    fn has_role(role: felt252, account: ContractAddress) -> bool;

    /// Returns the adming role that controls `role`.
    fn get_role_admin(role: felt252) -> felt252;

    /// Grants `role` to `account`.
    fn grant_role(role: felt252, account: ContractAddress);

    /// Revokes `role` from `account`.
    fn revoke_role(role: felt252, account: ContractAddress);

    /// Revokes `role` from self.
    fn renounce_role(role: felt252, account: ContractAddress);
}
----

`AccessControl` also lets you `renounce_role` from the calling account.
The method expects an account as input as an extra security measure, to ensure you are

> not renuncing a role from an unintended account.

Reviewer: ```suggestion
not renouncing a role from an unintended account.
```

> :ownable-cairo: link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/access/ownable/ownable.cairo[Ownable]

Reviewer 2: why not a release? `cairo-2` is likely to be deleted this week

Author: Should I replace it with `release-0.7.0` ?

Reviewer 2: `release-v0.7.0`, yes

The implementing contract's constructor should set the initial owner by passing the owner's address to Ownable's
xref:/api/access.adoc#AccessControl-initializer[`initializer`] like this:

[,javascript]
----
use openzeppelin::access::ownable::Ownable;

#[starknet::contract]
mod Contract {
    use starknet::ContractAddress;
    use super::Ownable;

    #[storage]
    struct Storage {}

    #[constructor]
    fn constructor(self: @ContractState, owner: ContractAddress) {
        let mut unsafe_state = Ownable::unsafe_new_contract_state();
        Ownable::InternalImpl::initializer(ref unsafe_state, owner);
    }

    (...)
}
----

To restrict a function's access to the owner only, add in the `assert_only_owner` method:

[,javascript]
----
#[starknet::contract]
mod Contract {
    use openzeppelin::access::ownable::Ownable;

    (...)

    #[external(v0)]
    fn foo(ref self: ContractState) {
        // This function can only be called by the owner
        let unsafe_state = Ownable::unsafe_new_contract_state();
        Ownable::InternalImpl::assert_only_owner(@unsafe_state);

        (...)
    }
}
----

=== Interface

> This is the full interface a contract must implement to be compliant with the `Ownable` implementation:

Reviewer 2: ```suggestion
This is the full interface of the `Ownable` implementation:
```
implementations are not specs, i insist it doesn't make sense to say something is "compliant with the implementation"

OpenZeppelin Contracts for Cairo provides {ownable-cairo} for implementing ownership in your contracts.

=== Usage

Integrating this module into a contract first requires assigning an owner.
The implementing contract's constructor should set the initial owner by passing the owner's address to Ownable's
`initializer` like this:

[,javascript]
----
use openzeppelin::access::ownable::Ownable;

#[starknet::contract]
mod Contract {
    use starknet::ContractAddress;
    use super::Ownable;

    #[storage]
    struct Storage {}

    #[constructor]
    fn constructor(self: @ContractState, owner: ContractAddress) {
        let mut unsafe_state = Ownable::unsafe_new_contract_state();
        Ownable::InternalImpl::initializer(ref unsafe_state, owner);
    }

    (...)
}
----

To restrict a function's access to the owner only, add in the `assert_only_owner` method:

[,javascript]
----

use openzeppelin::access::ownable::Ownable;

#[starknet::contract]
mod Contract {
    (...)

    #[external(v0)]
    fn foo(ref self: ContractState) {
        // This function can only be called by the owner
        self.assert_only_owner();

        (...)
    }
}
----

> === Interface

Author: I think presenting this here is worthwhile, and we present interfaces in other modules as well besides API.

This is the full interface a contract must implement to be compliant with the `Ownable` implementation:

[,javascript]
----
trait IOwnable {
    /// Returns the current owner.
    fn owner() -> ContractAddress;

    /// Transfers the ownership from the current owner to a new owner.
    fn transfer_ownership(new_owner: ContractAddress);

    /// Renounces the ownership of the contract.
    fn renounce_ownership();
}
----

Ownable also lets you:

- `transfer_ownership` from the owner account to a new one, and

> - `renounce_ownership` for the owner to relinquish this administrative privilege, a common pattern

Author: Are you suggesting to add a note? I don't think is necessary for what is worth. To the contract, DAOs are still a single owner, as any other contract could be. To the Ownable module, is not relevant what mechanism are you using behind the owner.

> - `renounce_ownership` for the owner to relinquish this administrative privilege, a common pattern

Reviewer 2: sorry i meant the line below, that talks about centralization. it was just a comment though, no need to fix anything.

Separately, you will be able to define rules for how accounts can be granted a role, have it revoked, and more.

Most software uses access control systems that are role-based: some users are regular users, some may be supervisors
or managers, and a few will often have administrative privileges.

=== Usage

For each role that you want to define, you will create a new _role identifier_ that is used to grant, revoke, and
check if an account has that role. See xref:#creating_role_identifiers[Creating role identifiers] for information
on creating identifiers.

Here's a simple example of implementing `AccessControl` on a portion of an ERC20 token contract which defines
and sets a 'minter' role:

[,javascript]
----
// sn_keccak('MINTER_ROLE')
const MINTER_ROLE: felt252 =
    0x32df0fed2c77648de5860a4cc508cd0818c85b8b8a1ab4ceeef8d981c8956a6;

#[starknet::contract]
mod Contract {
    use openzeppelin::access::accesscontrol::AccessControl::InternalImpl::assert_only_role;
    use openzeppelin::access::accesscontrol::AccessControl;
    use openzeppelin::token::erc20::ERC20;
    use starknet::ContractAddress;

>     use super::MINTER_ROLE;

Author: I think is easier to notice at the top, and the tutorial is focusing on these roles, I would let it on top if you agree.

        assert_only_role(@access_state, BURNER_ROLE);

        let mut erc20_state = AccessControl::unsafe_new_contract_state();
        ERC20::InternalImpl::_burn(ref erc20_state, account, amount);
    }
}
----

So clean!
By splitting concerns this way, more granular levels of permission may be implemented than were possible with the
simpler ownership approach to access control. Limiting what each component of a system is able to do is known
as the https://en.wikipedia.org/wiki/Principle_of_least_privilege[principle of least privilege], and is a good
security practice. Note that each account may still have more than one role, if so desired.

=== Granting and revoking roles

The ERC20 token example above uses `_grant_role`, an `internal` function that is useful when programmatically assigning
roles (such as during construction). But what if we later want to grant the 'minter' role to additional accounts?

By default, *accounts with a role cannot grant it or revoke it from other accounts*: all having a role does is making
the `assert_only_role` check pass. To grant and revoke roles dynamically, you will need help from the role's _admin_.

> Every role has an associated admin role, which grants permission to call the `grant_role` and `revoke_role` functions.

Reviewer 2: API links here too would be an improvement 🙇

A role can be granted or revoked by using these if the calling account has the corresponding admin role.
Multiple roles may have the same admin role to make management easier.
A role's admin can even be the same role itself, which would cause accounts with that role to be able
to also grant and revoke it.

This mechanism can be used to create complex permissioning structures resembling organizational charts, but it also
provides an easy way to manage simpler applications. `AccessControl` includes a special role with the role identifier
of `0`, called `DEFAULT_ADMIN_ROLE`, which acts as the *default admin role for all roles*.
An account with this role will be able to manage any other role, unless `_set_role_admin` is used to select a new admin role.

Let's take a look at the ERC20 token example, this time taking advantage of the default admin role:

[,javascript]
----
// sn_keccak('MINTER_ROLE')
const MINTER_ROLE: felt252 =
    0x32df0fed2c77648de5860a4cc508cd0818c85b8b8a1ab4ceeef8d981c8956a6;
// sn_keccak('BURNER_ROLE')
const BURNER_ROLE: felt252 =

>     0x11d16cbaffd01df69ce1c404f6340ee057498f5f00246190ea54220576a848;

Reviewer 2: replace with `selector!` please

same account can call `grant_role` to give minting or burning permission, and `revoke_role` to remove it.

Dynamic role allocation is often a desirable property, for example in systems where trust in a participant may vary
over time. It can also be used to support use cases such as https://en.wikipedia.org/wiki/Know_your_customer[KYC],
where the list of role-bearers may not be known up-front, or may be prohibitively expensive to include in a single transaction.

=== Creating role identifiers

In the Solidity implementation of AccessControl, contracts generally refer to the
https://docs.soliditylang.org/en/latest/units-and-global-variables.html?highlight=keccak256#mathematical-and-cryptographic-functions[keccak256 hash]
of a role as the role identifier.

For example:

[,javascript]
----
bytes32 public constant SOME_ROLE = keccak256("SOME_ROLE")
----

These identifiers take up 32 bytes (256 bits).

Cairo field elements (`felt252`) store a maximum of 252 bits.
With this discrepancy, this library maintains an agnostic stance on how contracts should create identifiers.
Some ideas to consider:

> * Use the {sn_keccak} instead.

Reviewer 2: ```suggestion
* Use {sn_keccak} instead.
```

* Use Cairo friendly hashing algorithms like Poseidon, which are implemented in the
https://github.com/starkware-libs/cairo/blob/main/corelib/src/poseidon.cairo[Cairo corelib].

> TIP: The `selector!` macro can be used for computing the {sn_keccak} in Cairo.

Reviewer 2: ```suggestion
TIP: The `selector!` macro can be used to compute {sn_keccak} in Cairo.
```

=== Interface

> This is the full interface a contract must implement to be compliant with the `AccessControl` implementation:

Reviewer 2: what does it mean to be compliant with the impl? isn't with the standard? although we don't mention any standard for this.

i think it's especially confusing to have a single function described below (`renounce_role`).

Author: > what does it mean to be compliant with the impl? isn't with the standard?

With our implementation of AccessControl, I think using the word standard without a standard will be confusing.

> i think it's especially confusing to have a single function described below (renounce_role).

The guide already explained the rest of the methods at this point, the only one pending was `renounce_role`, after following the tutorial it shouldn't be that confusing.

Reviewer 2: sure but there's a single element list. maybe it can be turned into a paragraph
![image](https://github.com/OpenZeppelin/cairo-contracts/assets/1703831/6075f8df-7ec6-4a7e-a23f-75ff417748fa)

> With our implementation of AccessControl, I think using the word standard without a standard will be confusing.

[there is](https://eips.ethereum.org/EIPS/eip-5982)

Author: > [there is](https://eips.ethereum.org/EIPS/eip-5982)

I didn't know there was a SRC for that already, nice.

Reviewer 2: no it's not a SRC but an ERC

Reviewer 2: ```suggestion
This is the full interface of the `AccessControl` implementation:
```
in line with the Ownable suggestion
